docs/grfstrip.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" Please adjust this date whenever revising the manpage.
.Dd October 24, 2017
.Dt GRFStrip 1
.Sh NAME
.Nm GRFID
.Nd A tool strip (some) real sprites from a GRF
.Sh SYNOPSIS
.Op Fl hv
.Nm grfstrip Ar origin dest (depth zoom)*
.Sh DESCRIPTION
GRFStrip strips all real sprites from a GRF and leaves only the ones
specified in the tuples at the command line and the 8bpp normal zoom.
The new GRF file will be written to dest.
.Sh OPTIONS
.Bl -tag
.It Fl h
Display the help message.
.It Fl v
Display the version.
.El
.Sh KNOWN SPRITE FORMATS
Known depths are: 8bpp, 32bpp
.Pp
Known zoom levels are: normal, zi4, zi2, zo2, zo4, zo8
.Sh SEE ALSO
The documentation in
.Pa /usr/share/doc/grfcodec/
.Sh AUTHOR
GRFStrip was written by Remko Bijker.
.Pp
This manual page was written by Remko Bijker.
.Sh COPYRIGHT
GRFStrip is Copyright (C) 2012 by Remko Bijker.


Changelog
Changes in
- (devel) Detect UPX, fix some race issues in the Makefile, disable
  optimization if GCC 4.5.0 is detected (to work around an optimization bug),
  fix version.h generation if there is a version.h in /usr/include.
- (devel) Fix a number of compiler warnings.
- Clamp the arguments for -w and -h to 65535 as that is the maximum size of a
  PCX.
- Silence progress-meter output when stdout is not a console.
- Unify warnings and errors so they go to stderr and are prefixed by "Warning:"
  or "Error:".
- Allow hex escapes that have bit 31 set.
- After cropping a sprite, compress and write the correct number of bytes.
- Add -X to print PCX sprite numbers in hex
- (devel) Many compilation issues with different compilers and architectures.
- Fix bug in UTF-8 generation for escapes \U800..\U7FFF.
- Fix U+E0xx characters being output quoted.
- (devel) By Tron: code cleanup, make C functions and variables static where
  possible, remove dead and debugging code.
- Require two slashes before dismissing anything as a C++ style comment.
- Fix reading of real sprites with Unixy absolute paths.
- Check for pixels with palette index of 255 when encoding.
- Add -q to silence warnings.
- Add -c to automatically crop extraenous transparent blue pixels.
- Add escapes for the new action 2 operations 0E, 0F, 11, 12, 13, 14, 15, 16
- Prevent quotes and escapes in sprite 0.
- Add escapes for long-format intro dates.

Changes in 0.9.10 (r819)
- (bugfix) Correct reading of \7c, \7C, \D| and \D&. (DaleStan)

Changes in 0.9.9 (r653)
- By Patchman:
  - Check for sprites that extend beyond the end of the PCX file.
  - Update versioning system to show last-comitted revision for non-release
    builds.
- By DaleStan:
  - (devel) Add release targets to the Makefiles
  - Check the palette of the PCX file before encoding from it.
    (Override a failed check with -f)
  - Introduce several backslash escape sequences; see the wiki for further
    information.
    (disable with -x or -xx)
  - Do not apply colormap to sprites that like character glyphs.
    (Apply colormap to all sprites with -M instead of -m)
  - Enable reading of GIMP palette files with -p.
  - (devel) Terse build mode; get the verbose mode with make V=1.
  - (devel) Touch version.h if and only if it changes.

Changes in 0.9.8 (r197)
- By DaleStan:
  - (devel) grfcodec (but not grfdiff or grfmerge) can now be built in
    MS Visual Studio 2003.
  - Slightly change the plaintext quoting algorithm.
  - (devel) Change several memcpy()s to pointer assignments.
  - (devel) Plug several memory leaks
  - (devel) Source code and Makefile changes to grfdiff and grfmerge build
    with both Makefiles, on both Cygwin and Linux.
  - Make code gcc 4 -Wall clean.
  - Use NFORenum's NFO reading engine:
    - Sprite numbers are now optional, and need not be correct if present.
    - Sprite lengths need not be correct. Set to 0 to supress the incorrect
      length warnings.
    - Real sprites are checked for format compliance.
  - Make 0-byte pseudo sprites an error.
  - (bugfix) Prevent binary includes from including directories on OSes that
    would othewise allow it.
  - (bugfix) It is now possible for a real sprite to start above the
    preceeding sprite (the infamous decreasing ypos bug)
  - grfcodec now exits with 0 on success.

Changes in 0.9.7
- support including binary files using "  4 ** data/dingdong.wav" syntax
- by Oskar: support // and # comments within pseudo-sprites
- fixed incorrect error message while encoding if .nfo file could not be read
  (now it reports "Cannot read %s" instead of Creating sprites/: File exists")
- made code -Wall clean
- remove myalloc stuff

Changes in 0.9.6
- by pasky:
  * made it possible to compile grfcodec on Linux
  * made compatible with gcc 3.x
  * when decoding, text strings are written in plain text (disable with -t)
  * supports plain text in .nfo files
  * fixed bug when first encoding a grf file wouldn't rename .new to .grf

Changes in 0.9.5
- self-extracting executables now work on Win2K etc., which doesn't pass
  the full path to the C library
- GRFMerge will just skip .grf files it can't find
- making sprites larger doesn't work if they're at the right edge; complains
  that ofsx too large; instead just use background colour
- GRFDiff doesn't like grfdiff -l 1254 ... (single value in -l), complains
  that there were no differences and makes no output
- allow GRFMerge to read GRFDiff's .exe files, both for -l and other options

Changes in 0.9.4
- fixed yet another crash while compressing
- sprites no longer need correct numbers in the .NFO file; just use "-1" for
  each number and it will be assumed to be correct

Changes in 0.9.3
- fixed crash due to buffer overrun while compressing
- slightly improved compression efficiency and speed
- GRFCodec now handles spaces in file/directory names properly
- GRFCodec now supports comments in .NFO files (from //, ; or # to end of line)
- speed up PCX processing

Changes in 0.9.2
- fixed crash if pcx file not found
- fixed crash if nfo file has no version info header

Changes in 0.9.1
- bug fix by Marcin: wrong order of test in encode function

Changes in release April 20 2002
- -p now actually uses the parameter you give it

Changes in release Dec 03 2001
- made -p option a valid option
- improve compression speed even more using assembly optimizations
- compression is now on by default, disable with -c-
- fixed bug where encoding would fail if the grf file didn't exist already
- fixed bug where decoding would fail creating the sprites directory
- corrected some mistakes in grf.txt (checksum, and initial offsets for tiles)

Changes in release March 26 2001
- fix bug that didn't flush last row if it only contained a single sprite
- fix bug with no height given (would make PCX have height -1)
- show percentage done while decoding
- fix compression algorithm
- improve compression performance
- try without compression if data is inconsistent after decompressing our

src/grfstrip.cpp
/*
 * This file is part of GRFCodec.
 * GRFCodec is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
 * GRFCodec is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

#include "md5.h"

#ifndef WIN32
#include <sys/mman.h>
#else
#include <stdlib.h>
#define PROT_READ 0
#define MAP_PRIVATE 0
/* Lets fake mmap for Windows, please! */
void *mmap (void *ptr, long size, long prot, long type, long handle, long arg) {
	char *mem = (char*)malloc(size + 1);
	mem[size] = 0;
	FILE *in = fdopen(handle, "rb");
	if (fread(mem, size, 1, in) != 1) {
		fclose(in);
		free(mem);
		return NULL;
	}
	fclose(in);
	return mem;
}
long munmap (void *ptr, long size) {
	free(ptr);
	return 0;
}
#endif

#include "version.h"

size_t _file_length;
uint8_t *_file_buffer;
uint8_t *_buffer;

inline void SkipBytes(size_t count)
{
	_buffer += count;
}

inline uint8_t ReadByte()
{
	return *(_buffer++);
}

inline uint16_t ReadWord()
{
	uint16_t v = ReadByte();
	return v | (ReadByte() << 8);
}

inline uint32_t ReadDWord()
{
	uint32_t v = ReadWord();
	return v | (ReadWord() << 16);
}

static const char header[] = {
	'\x00', '\x00',                 // End-of-file marker for old OTTDp versions
	'G',    'R',    'F',    '\x82', // Container version 2
	'\x0D', '\x0A', '\x1A', '\x0A', // Detect garbled transmission
};


const char *Strip(const char *origin, const char *dest, uint32_t allowed)
{
	const char *ret = NULL;
	uint32_t len;
	uint8_t *end;
	FILE *fout;
	FILE *fin;
	
	fin = fopen(origin, "rb");
	if (fin == NULL) {
		ret = "Unable to open origin file";
		goto fail_ret;
	}

	fout = fopen(dest, "wb");
	if (fout == NULL) {
		ret = "Unable to open destination file";
		goto fail_close_fin;
	}

	/* Get the length of the file */
	fseek(fin, 0, SEEK_END);
	_file_length = ftell(fin);
	fseek(fin, 0, SEEK_SET);

	/* Map the file into memory */
	_file_buffer = (uint8_t*)mmap(NULL, _file_length, PROT_READ, MAP_PRIVATE, fileno(fin), 0);
	end = _file_buffer + _file_length;

	if (_file_length <= sizeof(header) || memcmp(_file_buffer, header, sizeof(header)) != 0) {
		return "No GRF with container version 2.";
		goto fail;
	}

	_buffer = _file_buffer + sizeof(header);
	len = ReadDWord();
	_buffer = _file_buffer + sizeof(header) + len + 4;

	if (_buffer >= end) {
		return "Invalid GRF";
		goto fail;
	}

	if (fwrite(_file_buffer, 1, _buffer - _file_buffer, fout) != (size_t)(_buffer - _file_buffer)) {
		ret = "Could not write to file";
		goto fail;
	}

	for (;;) {
		uint8_t *begin = _buffer;
		uint32_t id = ReadDWord();
		if (id == 0) {
			if (fwrite(begin, 1, end - begin, fout) != (size_t)(end - begin)) {
				ret = "Could not write to file";
				goto fail;
			}
			break;
		}

		uint32_t size = ReadDWord();
		uint8_t info = ReadByte();
		uint8_t zoom = ReadByte();

		int offset = ((info & 0x7) == 4 ? 0 : 16) + zoom;

		if (info == 0xFF || (allowed & (1 << offset)) != 0) {
			/* Copy */
			if (fwrite(begin, 1, size + 8, fout) != (size_t)(size + 8)) {
				ret = "Could not write to file";
				goto fail;
			}
		}

		/* Skip sprite. */
		SkipBytes(size - 2);
		if (_buffer >= end) { 
			ret = "Invalid GRF";
			goto fail;
		}
	}

fail:
	munmap(_file_buffer, _file_length);
	fclose(fout);
fail_close_fin:
	fclose(fin);
fail_ret:
	return ret;
}

int main(int argc, char **argv)
{
	static const char * const depth[] = { "8bpp", "32bpp" };
	static const int num_depths = sizeof(depth) / sizeof(depth[0]);
	static const char * const zoom[] = { "normal", "zi4", "zi2", "zo2", "zo4", "zo8" };
	static const int num_zooms = sizeof(zoom) / sizeof(zoom[0]);

	if (argc < 3 || strcmp(argv[1], "-h") == 0) {
		printf(
			"GRFSTRIP " VERSION " - Copyright (C) 2009 by Remko Bijker\n"
			"\n"
			"Usage:\n"
			"    GRFSTRIP <origin> <dest> (<depth> <zoom>)*\n"
			"        Strip real sprites that are not in the set \"normal 8bpp and the ones\n"
			"        specified at the command line\" from origin into dest.\n"
			"        Known depths: ");
		for (int j = 0; j < num_depths; j++) {
			if (j != 0) printf(", ");
			printf("%s", depth[j]);
		}
		printf("\n        Known zooms: ");
		for (int j = 0; j < num_zooms; j++) {
			if (j != 0) printf(", ");
			printf("%s", zoom[j]);
		}
		printf( "\n"
			"    GRFSTRIP -v\n"
			"        Get the version of GRFSTRIP\n"
			"\n"
			"You may copy and redistribute it under the terms of the GNU General Public\n"
			"License, as stated in the file 'COPYING'.\n");
		return 1;
	}
	if (strcmp(argv[1], "-v") == 0) {
		printf("GRFSTRIP " VERSION " - Copyright (C) 2012 by Remko Bijker\n");
		return 0;
	}

	uint32_t allowed = 1;
	for (int i = 3; i + 1 < argc; ) {
		int depth_offset = 0xFF;
		for (int j = 0; j < num_depths; j++) {
			if (strcmp(argv[i], depth[j]) == 0) {
				depth_offset = 16 * j;
				break;
			}
		}
		if (depth_offset == 0xFF)  {
			printf("Invalid depth \"%s\"\n", argv[i]);
			return 1;
		}
		i++;

		int zoom_offset = 0xFF;
		for (int j = 0; j < num_zooms; j++) {
			if (strcmp(argv[i], zoom[j]) == 0) {
				zoom_offset = j;
				break;
			}
		}
		if (zoom_offset == 0xFF) {
			printf("Invalid zoom \"%s\"\n", argv[i]);
			return 1;
		}
		i++;

		allowed |= 1 << (depth_offset + zoom_offset);
	}

	const char *err = Strip(argv[1], argv[2], allowed);
	if (err == NULL) {
		printf("Stripped %s into %s successfully\n", argv[1], argv[2]);
		return 0;
	}

	printf("Unable to get requested information: %s\n", err);
	return 1;
}

